nginx/modsecurity/ngx_http_modsecurity.c

```diff
@@ -848,19 +898,81 @@ ngx_http_modsecurity_body_handler(ngx_http_request_t *r)
     ngx_http_core_run_phases(r);
 }
 
+static ngx_int_t
+ngx_http_calculate_modsec_latency(ngx_http_request_t *r, ngx_http_modsecurity_ctx_t* ctx,
+                                  ngx_http_modsecurity_loc_conf_t * cf)
+{
+    ngx_time_t *end_time;
+    ngx_msec_int_t   ms;
+    ngx_int_t ret;
+    ngx_time_update();
+    end_time = ngx_timeofday();
+
+    ms = (ngx_msec_int_t)
+        ((end_time->sec - ctx->start_time->sec) * 1000 + (end_time->msec - ctx->start_time->msec));
+    ms = ngx_max(ms, 0);
+
+    ngx_http_variable_value_t* modsec_latency_detect_var = ngx_http_get_indexed_variable(r, modsec_latency_detect_index);
+    ngx_http_variable_value_t* modsec_latency_prevent_var = ngx_http_get_indexed_variable(r, modsec_latency_prevent_index);
+    if (cf->config->is_enabled == MODSEC_DETECTION_ONLY) {
+        ret = ngx_http_modsecurity_set_modsec_latency(r, modsec_latency_detect_var, ms);
+        if (ret != NGX_OK) {
+            return ret;
+        }
+        return ngx_http_modsecurity_set_modsec_latency(r, modsec_latency_prevent_var, 0);
+    }
+    else {
+        ret = ngx_http_modsecurity_set_modsec_latency(r, modsec_latency_prevent_var, ms);
+        if (ret != NGX_OK) {
+            return ret;
+        }
+        return ngx_http_modsecurity_set_modsec_latency(r, modsec_latency_detect_var, 0);
+    }
+}
 
 /*
 ** [ENTRY POINT] does : this function called by nginx from the request handler
+*   Calculate the modsec latency in this function and invoke the main handler.
 */
 static ngx_int_t
-ngx_http_modsecurity_handler(ngx_http_request_t *r)
+ngx_http_modsecurity_handler_with_timer(ngx_http_request_t *r)
 {
+    ngx_time_update();
```

Comment:
Move this after the null check on 947. We only need to update the time if we have ctx == null.

Reply:
done